qiskit/qpy/binary_io/circuits.py

```diff
@@ -1070,12 +1121,55 @@ def _write_pauli_evolution_gate(file_obj, evolution_gate, version):
     def _write_elem(buffer, op):
         elem_data = common.data_to_binary(op.to_list(array=True), np.save)
         elem_metadata = struct.pack(formats.SPARSE_PAULI_OP_LIST_ELEM_PACK, len(elem_data))
+        if version >= 17:
+            elem_sparse_operator = struct.pack("?", False)
+            buffer.write(elem_sparse_operator)
         buffer.write(elem_metadata)
         buffer.write(elem_data)
 
+    def _write_elem_sparse(buffer, op):
+        bitterms = op.bit_terms
+        coeffs = op.coeffs
+        bounds = op.boundaries
+        inds = op.indices
+        numq = op.num_qubits
+
+        coeff_real = []
+        coeff_imag = []
+        for coeff in coeffs:
+            coeff_real.append(coeff.real)
+            coeff_imag.append(coeff.imag)
+        coeff_data = struct.pack(
+            f"{len(coeffs)*2}d", *(val for pair in zip(coeff_real, coeff_imag) for val in pair)
+        )
+        bitterm_data = struct.pack(f"{len(bitterms)}i", *bitterms)
```

Comment:
Could we optimize the data types we're using here? 
* bit terms are number up to 11 at most, so a `unsigned short` should be fine here (`"h"`)
* indices are unsigned so we could use `"I"`
* bounds should be `"Q"` to be safe, since there's no limit on the number of terms we can have, but they surely are unsigned